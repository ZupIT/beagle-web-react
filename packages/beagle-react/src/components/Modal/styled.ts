import styled from 'styled-components'
import { BeagleTheme } from '../commons.styled'

export const StyledModal = styled.div`
  background: ${BeagleTheme.darkGray};
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 1;
  display: flex;
  align-items: center;
  justify-content: center;

  > div {
    background-color: #FFF;
    border-radius: 25px;
    padding: 30px;
  }
`
