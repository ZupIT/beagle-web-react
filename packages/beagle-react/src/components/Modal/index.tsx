/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import React, { FC, useEffect, useRef } from 'react'
import { logger } from '@zup-it/beagle-web'
import { BeagleModalInterface } from 'common/models'
import { buildAccessibility } from '../../../../common/utils/accessibility'
import withTheme from '../utils/withTheme'
import { StyledModal } from './styled'

/**
 * @deprecated since version 1.8.0.
 * This component will be removed in a future version.
*/
const Modal: FC<BeagleModalInterface> = props => {
  const { isOpen, onClose, style, className, children, accessibility } = props
  const elementRef = useRef<HTMLDivElement>(null)
  const a11y = buildAccessibility(accessibility)

  function closeOnEsc({ key }: KeyboardEvent) {
    if (key === 'Escape') {
      onClose()
      document.removeEventListener('keyup', onClose)
    }
  }

  function closeOnClickOutside({ target }: React.MouseEvent<HTMLDivElement>) {
    if (!elementRef.current || !elementRef.current.firstChild) return
    if (!elementRef.current.firstChild.contains(target as Node)) onClose()
  }

  useEffect(() => {
    document.addEventListener('keyup', closeOnEsc)
    logger.warn('This component is deprecated since version 1.8.0 and will be removed in a future version.')
    return () => document.removeEventListener('keyup', closeOnEsc)
  }, [])

  const modal = (
    <StyledModal 
      ref={elementRef}
      style={style}
      className={className}
      onClick={closeOnClickOutside}
      aria-modal={true}
      {...a11y}>
      {children}
    </StyledModal>
  )

  return isOpen ? modal : null
}

export default withTheme(Modal)
