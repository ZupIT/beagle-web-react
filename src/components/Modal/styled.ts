import styled from 'styled-components'

export const StyledModal = styled.div`
  background: rgba(0, 0, 0, 0.6);
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;

  > div {
    background-color: #FFF;
    border-radius: 25px;
    width: 800px;
    padding: 20px;
  }
`
