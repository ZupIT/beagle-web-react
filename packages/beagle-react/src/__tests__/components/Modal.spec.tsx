/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

// Link.react.test.js
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import { configure, mount } from 'enzyme'
import Modal from '../../components/Modal'

configure({ adapter: new Adapter() })

test('Beagle snapshot Modal', () => {
  const onClickMock = jest.fn()
  let wrapper =  mount(<Modal onClose={onClickMock} isOpen={true} />)

  expect(wrapper).toMatchSnapshot()
})

test('Should render an opened modal', () => {
  let wrapperOpened = mount(
    <Modal onClose={jest.fn()} isOpen={true}>
      Opened modal
    </Modal>
  )

  expect(wrapperOpened.text()).toContain('Opened modal')
})

test('Should render a closed modal', () => {
  let wrapperClosed = mount(
    <Modal onClose={jest.fn()} isOpen={false}>
      Closed modal
    </Modal>
  )

  expect(wrapperClosed.text()).not.toContain('Closed modal')
})

test('Should render an open modal with the children inside, then close, then open again', () => {
  let wrapperOpened = mount(
    <Modal onClose={jest.fn()} isOpen={true}>
      Opened modal
    </Modal>
  )

  expect(wrapperOpened.text()).toContain('Opened modal')
  wrapperOpened.setProps({ isOpen: false })
  expect(wrapperOpened.text()).not.toContain('Opened modal')
  wrapperOpened.setProps({ isOpen: true })
  expect(wrapperOpened.text()).toContain('Opened modal')
})
