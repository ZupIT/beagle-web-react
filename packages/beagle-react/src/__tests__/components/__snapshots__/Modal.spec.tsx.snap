// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle snapshot Modal 1`] = `
<ComponentWithTheme
  isOpen={true}
  onClose={[MockFunction]}
>
  <Modal
    className=""
    isOpen={true}
    onClose={[MockFunction]}
  >
    <styled.div
      className=""
      onClick={[Function]}
    >
      <div
        className="sc-bdfBwQ eBKHls"
        onClick={[Function]}
      />
    </styled.div>
  </Modal>
</ComponentWithTheme>
`;
